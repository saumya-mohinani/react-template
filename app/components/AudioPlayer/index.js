import React, { useState, useRef } from 'react';
import { PlayCircleFilled, PauseCircleFilled } from '@ant-design/icons';
import { Tooltip } from 'antd';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';

const PlayIcon = styled(PlayCircleFilled)`
  && {
    font-size: 250%;
  }
`;
const PauseIcon = styled(PauseCircleFilled)`
  && {
    font-size: 250%;
  }
`;
const Timeline = styled.div`
  && {
    position: relative;
    width: 100%;
    height: 100%;
    background-color: #e3e3e3;
    border: 1px solid black;
  }
`;
const TimelineContainer = styled.div`
  && {
    position: relative;
    width: 100%;
    margin: 1em 5px;
    height: 5px;
  }
`;

const Progress = styled.div`
  && {
    position: absolute;
    left: 0;
    top: 0;
    height: 100%;
    background-color: coral;
  }
`;
const Controls = styled.div`
  && {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    margin-top: 1em;
  }
`;

const AudioPlayer = ({ source, intl }) => {
  const [playing, setPlaying] = useState(false);
  const audio = useRef(null);
  const timeline = useRef(null);
  const progress = useRef(null);
  const controls = useRef(null);
  const currentTime = useRef(null);

  const calculatePercentPlayed = () => {
    let percentPlayed = (audio.current.currentTime / audio.current.duration).toFixed(2) * 100;
    progress.current.style.width = `${percentPlayed}%`;
  };
  const calculateCurrentValue = currentTime => {
    const currentMinute = parseInt(currentTime / 60) % 60;
    const currentSecondsLong = currentTime % 60;
    const currentSeconds = currentSecondsLong.toFixed();
    const currentTimeFormatted = `${currentMinute < 10 ? `0${currentMinute}` : currentMinute}:${
      currentSeconds < 10 ? `0${currentSeconds}` : currentSeconds
    }`;

    return currentTimeFormatted;
  };
  const initProgressBar = () => {
    const media = audio.current;
    const time = calculateCurrentValue(media.currentTime);

    function seek(e) {
      e.stopPropagation();
      const percent = e.offsetX / timeline.current.offsetWidth;
      media.currentTime = percent * media.duration;
    }

    currentTime.current.innerHTML = time;
    timeline.current.addEventListener('click', seek);
    calculatePercentPlayed();
  };

  const onEnded = () => {
    progress.current.style.width = 0;
    currentTime.current.innerHTML = '00:00';

    setPlaying(false);
  };

  const play = e => {
    e.stopPropagation();
    setPlaying(true);
    audio.current.play();
  };

  const pause = e => {
    e.stopPropagation();
    setPlaying(false);
    audio.current.pause();
  };

  return (
    <div>
      <Controls ref={controls}>
        {playing ? (
          <Tooltip title={intl.formatMessage({ id: 'pause' })}>
            <PauseIcon onClick={e => pause(e)} />
          </Tooltip>
        ) : (
          <Tooltip title={intl.formatMessage({ id: 'play' })}>
            <PlayIcon onClick={e => play(e)} />
          </Tooltip>
        )}
        <TimelineContainer>
          <Timeline ref={timeline}>
            <Progress ref={progress}></Progress>
          </Timeline>
        </TimelineContainer>

        <small id="currentTime" ref={currentTime}>
          00:00
        </small>
      </Controls>
      <audio ref={audio} src={source} onTimeUpdate={initProgressBar} onEnded={onEnded} />
    </div>
  );
};

AudioPlayer.propTypes = {
  source: PropTypes.string.isRequired,
  intl: PropTypes.object
};

export default compose(injectIntl)(AudioPlayer);
