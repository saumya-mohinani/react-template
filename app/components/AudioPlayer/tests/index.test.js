/**
 *
 * Tests for LazyImage
 *
 */

import React from 'react';
import { renderWithIntl } from '@utils/testUtils';
import AudioPlayer from '../index';

describe('<AudioPlayer />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<AudioPlayer />);
    expect(baseElement).toMatchSnapshot();
  });
});
