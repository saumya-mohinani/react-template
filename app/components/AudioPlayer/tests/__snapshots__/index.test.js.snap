// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="AudioPlayer__Controls-vuu9fp-5 iAXODq"
      >
        <span
          aria-label="play-circle"
          class="anticon anticon-play-circle AudioPlayer__PlayIcon-vuu9fp-0 jBFxsf"
          role="img"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            data-icon="play-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm144.1 454.9L437.7 677.8a8.02 8.02 0 01-12.7-6.5V353.7a8 8 0 0112.7-6.5L656.1 506a7.9 7.9 0 010 12.9z"
            />
          </svg>
        </span>
        <div
          class="AudioPlayer__TimelineContainer-vuu9fp-3 hUDZYt"
        >
          <div
            class="AudioPlayer__Timeline-vuu9fp-2 eCUPHn"
          >
            <div
              class="AudioPlayer__Progress-vuu9fp-4 cBuoQg"
              style="width: 0%;"
            />
          </div>
        </div>
        <small
          id="currentTime"
        >
          00:00
        </small>
      </div>
      <audio />
    </div>
  </div>
</body>
`;
